04_Scheduling/03_Tasks_And_Threads.md

```diff
@@ -62,13 +62,109 @@ typedef struct
+
+```c
+typedef struct {
+    size_t pid;
+    char name[NAME_MAX_LEN];
+    status_t process_status;
+    cpu_status_t context;
+    uint64_t pdbr;
+    Heap_Node* heap_root;
+    Heap_Node* cur_heap_position;
+} process_t;
+
+```
+
+You can notie that we have actually added two variabiles, this is because of our memory allocation algorithm. 
+
+Of course we need to make changes to our memory allocation function, the logic can be kept the same but what we need to change are the variable used, now we no longer want to use a global variable as a base for our allocator, but we want to pick it from the current process. So If we take for example the third_alloc() function in the Heap chapter what we need to do is replace the references to the global variables with the pointer stored in the current process, so we will have something like: 
+
+```c
+void *third_alloc(size_t size) {
+  Heap_Node* cur_heap_position = get_current_running_process()->cur_heap_position;
+  cur_pointer = get_current_running_process()->heap_root;
+
+  while(cur_pointer < cur_heap_position) {
+    cur_size = *cur_pointer;
+    status = *(cur_pointer + 1);
+
+    if(cur_size >= size && status == FREE) {
+       status = USED;
+       return cur_pointer + 2;
+    }
+    cur_pointer = cur_pointer + (size + 2);
+  }
+
+  *cur_heap_position=size;  
+  cur_heap_position = cur_heap_position + 1;
+  *cur_heap_position = USED;
+  cur_heap_position = cur_heap_position + 1;
+  uint8_t *addr_to_return = cur_heap_position;
+  cur_heap_position+=size;
+  return (void*) addr_to_return;
+}
+```
+
+Be careful that the function above was not a complete one, we used it just as a nexample of how the updated code shoul look like, and of course we must do similar changes to the free function too. 
+
+What does it mean? Let's make an example: 
+
+* Process 1 make the following malloc call:
+
+```c
+    int *a = malloc(sizeof(int));
+```
+
+* Proess 2 make the following call:
+
+```c
+    custom_struct *cs = malloc(sizeof(custom_struct))
+```
+
+And in both cases the address returned is 0x10000. Apparently they are the same. But the pdb address in process 1 is different from pdbr address in process 2, and this means that we will have different page directories/and tables used. Since they are different we are physically referring to two different items, and they are not related so the will contain two different physical addresses (yes we can in theory having them pointing to the same physical address, and in some case we will also do it, for example when mapping the kernel, but is our physical memory manager duty to make sure that we don't allocate the same physical address twice). 
```

Comment:
Nice example!
We should add that this is probably not a good idea of how to do it in practice. For the kernel you will want a single heap, unless you can be certain of which process will be running when some memory is accessed. Quite often in the kernel you will want to access global state, meaning it must be visible to all processes, and therefore can't be stored on a per-process heap. There are of course exceptions to this, but I haven't come across enough for a per-process heap to be worth it.
For a user program they will provided their own heap (either custom, or the one part of the standard library they linked against). This operates in userspace so there isn't the massive overhead of a system call with every alloc/free call.